src/parser/validator.ts

```diff
@@ -0,0 +1,52 @@
+import { RootParser } from "../parser/RootParser";
+import { makeCstVisitor, VisitorResult, CommandNode } from "../parser/makeCstVisitor";
+
+interface validateCSTResult {
+    errors: ValidateCSTError[];
+    hasErrors: boolean;
+}
+
+interface ValidateCSTError {
+    error: Error,
+}
+
+export function validateVisitorResult(visitorResult: VisitorResult): validateCSTResult {
+    var validateResult = { errors: [], hasErrors: false};
+    var idSet = new Set<String>();
+    visitorResult.commands.forEach((command) => {
+        validateCST(command, validateResult, idSet);
+    });
+    return validateResult;
+}
+
+function validateCST(commandNode: CommandNode, errors: validateCSTResult, idSet: Set<String>) {
+    commandNode.children.forEach((child) => {
+        validateCST(child, errors, idSet);
+    });
+    codeBlockHasID(commandNode, errors);
+    idIsUnique(commandNode, errors, idSet);
+    return {
+        errors: errors,
+        hasErrors: errors.errors.length > 0
```

Comment:
The fact that we have `errors.errors...` here indicates a naming or data modeling issue somewhere.